core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,589 @@
+    decode = D.decodeBytes
+
+instance Bi Text.Text where
+    encode = E.encodeString
+    decode = D.decodeString
+
+instance Bi BS.Lazy.ByteString where
+    encode = encode . BS.Lazy.toStrict
+    decode = BS.Lazy.fromStrict <$> decode
+
+instance Bi a => Bi [a] where
+    encode = encodeList
+    decode = decodeList
+
+instance (Bi a, Bi b) => Bi (Either a b) where
+    encode (Left  x) = E.encodeListLen 2 <> E.encodeWord 0 <> encode x
+    encode (Right x) = E.encodeListLen 2 <> E.encodeWord 1 <> encode x
+
+    decode = do D.decodeListLenOf 2
+                t <- D.decodeWord
+                case t of
+                  0 -> do !x <- decode
+                          return (Left x)
+                  1 -> do !x <- decode
+                          return (Right x)
+                  _ -> fail $ "decode@Either: unknown tag " <> show t
+
+instance Bi a => Bi (NonEmpty a) where
+  encode = defaultEncodeList . toList
+  decode = do
+    l <- defaultDecodeList
+    case nonEmpty l of
+      Nothing -> fail "Expected a NonEmpty list, but an empty list was found!"
+      Just xs -> return xs
+
+instance Bi a => Bi (Maybe a) where
+    encode Nothing  = E.encodeListLen 0
+    encode (Just x) = E.encodeListLen 1 <> encode x
+
+    decode = do n <- D.decodeListLen
+                case n of
+                  0 -> return Nothing
+                  1 -> do !x <- decode
+                          return (Just x)
+                  _ -> fail $ "decode@Maybe: unknown tag " <> show n
+
+encodeContainerSkel :: (Word -> E.Encoding)
+                    -> (container -> Int)
+                    -> (accumFunc -> E.Encoding -> container -> E.Encoding)
+                    -> accumFunc
+                    -> container
+                    -> E.Encoding
+encodeContainerSkel encodeLen size foldr f  c =
+    encodeLen (fromIntegral (size c)) <> foldr f mempty c
+{-# INLINE encodeContainerSkel #-}
+
+decodeContainerSkelWithReplicate
+  :: Bi a
+  => D.Decoder s Int
+     -- ^ How to get the size of the container
+  -> (Int -> D.Decoder s a -> D.Decoder s container)
+     -- ^ replicateM for the container
+  -> ([container] -> container)
+     -- ^ concat for the container
+  -> D.Decoder s container
+decodeContainerSkelWithReplicate decodeLen replicateFun fromList = do
+    -- Look at how much data we have at the moment and use it as the limit for
+    -- the size of a single call to replicateFun. We don't want to use
+    -- replicateFun directly on the result of decodeLen since this might lead to
+    -- DOS attack (attacker providing a huge value for length). So if it's above
+    -- our limit, we'll do manual chunking and then combine the containers into
+    -- one.
+    size <- decodeLen
+    limit <- D.peekAvailable
+    if size <= limit
+       then replicateFun size decode
+       else do
+           -- Take the max of limit and a fixed chunk size (note: limit can be
+           -- 0). This basically means that the attacker can make us allocate a
```

Comment:
If I understand correctly, all this code is doing (Andrzej wrote it) is to pessimistically allocate at max 128 bytes as the size of a chunk if a potential attacker sent us a very big length (aka the `size` we decode with `decodeLen`) and `limit` is smaller than 128. As limit can be zero (we reached the capacity) an attacker could indeed make us allocate an extra 128 bytes, which yes, I think that's possible those won't fit in memory. I'm more concerned though about this line:

``` haskell
containers <- sequence $ buildOne m : replicate d (buildOne chunkSize)
```

I might be undercaffeinated, but `d` is the result of `size `divMod` chunkSize`, which can be quite big itself. An example:

``` haskell
> (maxBound :: Int) `divMod` 128
(72057594037927935,127)
```
This will then call `replicate d` and squash everything with `sequence` on a potentially very-large list, which sounds scary. Now, the question is how much big the buffer returned by `peekAvailable` is, as `chunkSize` is taking the max between it and `128`.

I think that, _in practice_, it's quite unrealistic that `limit` is going to be so small (it's probably going to be in the same order of magnitude of `size` as they are both capped by the `Int` size), but I will get back to you on this, just to be on the safe side!